### DISCUSSION TITLE
How to properly cache database results to improve `getStaticProps` and `getStaticPaths` results?

### DISCUSSION THREAD

**Discussion by aleksey-shmatov** [vercel/next.js#14533] [Help]
I have several pages which use a lot of common calls to database. `getStaticPaths` is basically exactly same for them and in `getStaticProps` also there is one exactly same database call per page. How can I make sure that database results are cached and reused when building these pages? I tried to add basic memoization but that seems like did nothing. My guess is that pages might be rendered in different workers and don't share memory space.

**Comment by ghost** [comment]
if it is for the fronted part its going to help you 
[SWR](https://swr.now.sh/) is the great option to  cache the data which will not fetch data again and again if your data is not changing go through this [video](https://www.youtube.com/watch?v=a7JigiLw_OY) you will have more clear idea

**Reply by aleksey-shmatov** [reply]
Unfortunately this is not helpful in my case. First issue is that I access database directly. Second issue is that it seems `SWR` also caches results in memory. As I mentioned I already tried caching in memory but it does not work, I suspect due to pages being rendered in different workers.

**Comment by revskill10** [comment]
One hacky solution is you write data to `/tmp` folder in `getStaticPaths`, then consume those data from `getStaticProps`

**Reply by aleksey-shmatov** [reply]
I am thinking about trying this but my main concern is that if page renders can happen in parallel database requests can still run in parallel(ie result from first one is not written to file yet when second issued). Technically I can add my own build step which would prepare files in advance to avoid this. 
Just wondering if there is common/recommended solution to this. Can't seem to find any relevant information in docs.

**Reply by isaachinman** [reply]
@aleksey-shmatov @revskill10 Did either of you end up finding a clean solution?

**Reply by aleksey-shmatov** [reply]
@isaachinman We switched to much faster database and problem became less relevant. I believe there is no easy way to achieve this.

**Reply by isaachinman** [reply]
I wrote a simple fs util to cache data. Should help speed up some build times:

```js
import fs from 'fs'
import path from 'path'

type CacheConfig = {
  cacheBaseKey: string
  ttl: number
}

const CACHE_DIR = path.join(process.cwd(), './.next-fs-cache')
const TTL_EXTENSION = '.ttl'

export const useFileSystemCache = (
  fn: (...args: any[]) => Promise<any>,
  { cacheBaseKey, ttl }: CacheConfig,
) => async (...args: any[]) => {
  const cacheKey = `${cacheBaseKey}${encodeURIComponent(JSON.stringify(args))}`

  const contentPath = path.join(CACHE_DIR, cacheKey)
  const ttlPath = `${contentPath}${TTL_EXTENSION}`

  const cacheMiss = async () => {
    const content = await fn(...args)
    fs.writeFileSync(contentPath, JSON.stringify(content))
    fs.writeFileSync(
      ttlPath,
      JSON.stringify(
        new Date(Date.now() + ttl)
      )
    )
    return content
  }

  if (!fs.existsSync(CACHE_DIR)) {
    fs.mkdirSync(CACHE_DIR)
  }

  if (fs.existsSync(contentPath) && fs.existsSync(ttlPath)) {
    const ttl = new Date(
      JSON.parse(
        fs.readFileSync(ttlPath, 'utf8')
      )
    )

    if (ttl.valueOf() > new Date().valueOf()) {
      return JSON.parse(
        fs.readFileSync(contentPath, 'utf8')
      )
    } else {
      return cacheMiss()
    }

  } else {
    return cacheMiss()
  }
}
```

**Reply by supagu** [reply]
Can you post an example of how to use this?
